{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Task 1: A simple dotplot program\n",
    "Overall aim: To build a program that creates a simple dotplot using no graphics but simply printing a matrix that indicates where two sequences show similarity.\n",
    "\n",
    "Your program should contain the following functions:\n",
    "\n",
    "    A) A function (called dotplot) that takes two sequences and fills a 2D matrix with ones and zeroes, according to whether a position corresponds to two identical residues or two different ones.\n",
    "\n",
    "    Arguments: two strings, one for each sequence\n",
    "    Returns: a 2D matrix filled in with zeroes and ones\n",
    "\n",
    "Note: You should write a separate function to create your own 2D matrix (call the function create_mat that takes two integer arguments and creates a matrix with number of columns and number of rows defined by the two arguments i.e. create_mat(nrwos, ncols))\n",
    "\n",
    "Improvement: It is generally accepted that the numpy package should be used for creating and manipulating matrices in python. Find out how to create a 2D matrix in numpy and change your code to replace the create_mat function with a create_mat_numpy function. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np \n",
    "\n",
    "def create_mat(nrows: int, ncols: int):\n",
    "    \"\"\"\n",
    "    Creates a blank matrix of zeroes\n",
    "\n",
    "    Args\n",
    "    nrows: Integer, number of rows\n",
    "    ncols: Integer, number of columns\n",
    "    \"\"\"\n",
    "    new_mat = np.zeros((nrows, ncols))\n",
    "    return new_mat\n",
    "\n",
    "def dotplot(seq1: str, seq2: str):\n",
    "    \"\"\"\n",
    "    Creates a dotplot from two sequences and returns a 2D matrix filled with ones and zeroes.\n",
    "\n",
    "    Args\n",
    "    seq1: String\n",
    "    seq2: String\n",
    "    \"\"\"\n",
    "    seq1_array = np.array(list(seq1))\n",
    "    seq2_array = np.array(list(seq2))\n",
    "    \n",
    "    dotplot_mat = create_mat(len(seq1), len(seq2))\n",
    "    \n",
    "    row_indices, col_indices = np.where(seq1_array[:, None] == seq2_array)\n",
    "    dotplot_mat[row_indices, col_indices] = 1\n",
    "\n",
    "    return dotplot_mat\n",
    "\n",
    "dotplot_dh = dotplot(\"DOROTHYHODGKIN\", \"DOROTHYCROWFOOTHODGKIN\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "    B) A function called print_dotplot that prints out the dotplot\n",
    "\n",
    "    Arguments: two strings representing the sequences, one 2D list representing the matrix already filled with zeroes and ones, and two characters to be used for printing out \"ones\" and \"zeroes\" (note: a common convention is to use an asterisk to indicate ones in the matrix and a white space to indicate zeroes)\n",
    "    Returns: nothing (simply prints the dotplot)\n",
    "\n",
    "Note: Use the function sys.stdout.write from the package sys to print to the standard out stream (screen in this case)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*                *    \n",
      " * *     *  **  *     \n",
      "  *     *             \n",
      " * *     *  **  *     \n",
      "    *         *       \n",
      "     *         *      \n",
      "      *               \n",
      "     *         *      \n",
      " * *     *  **  *     \n",
      "*                *    \n",
      "                  *   \n",
      "                   *  \n",
      "                    * \n",
      "                     *\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "\n",
    "def print_dotplot(dotplot_matrix, one = \"*\", zero = \" \"):\n",
    "    for row in dotplot_matrix:\n",
    "        for element in row:\n",
    "            if element == 0:\n",
    "                sys.stdout.write(str(zero))\n",
    "            else:\n",
    "                sys.stdout.write(str(one))\n",
    "        sys.stdout.write('\\n')\n",
    "    return\n",
    "\n",
    "print_dotplot(dotplot_dh)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dotplots can be represented as heatmaps (especially if real similarity values are used instead of just 0 and 1 values). Write a simple function to plot a heatmap using matplotlibs imshow function to plot your dotplot matrix as a heatmap.\n",
    "\n",
    "Arguments: a numpy 2D array containing the values of similarity between two sequences and two strings (one for each sequence)\n",
    "\n",
    "Returns: nothing.\n",
    "\n",
    "Improvement: Try to change the axes of the plot so that the ticks on each axis are labeled with the characters of each string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAW8AAAD3CAYAAADSftWOAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAELxJREFUeJzt3W2sZWV5xvH/5ZS3mSlSYNBEEEuBpMUawKFJSY0txAappqKBmJgSAinRYGwxJCakUfzAl6aDECEhWsOICak2oYBJY2uxRFGMZWBQpjFMBgq0tvKWQhEGZLj7Ye+jm+M5Z6/9svaZh/n/kpPMWmc/67n3nnWus87aa907VYUkqS1vWO8CJEmTM7wlqUGGtyQ1yPCWpAYZ3pLUIMNbkhq00PBOsi/JziS7kjyQ5JNJOtWQ5NgktyfZnWRPkuuSHDzF3A8m+XqSIyYYO9XcSY4azrkzyf8k+a+R5S7jn1+2fFGS6zvWnCR3J3nvyLoLknyj4/g3J/m74fP99yT/mOTkjmMX+lon+VySvxxZ/qckfzuyvC3JJyeoe+nrbX3WPaexU+9js+xfs9Q9Sw4Mx78pyS1JHk6yI8k9Sc7rOPb5kX+fO6z9rV3nXr6NCcZUkm0jy1ckuWrS7Yxa9JH3i1V1alWdArwHOBf4zLhBSQLcCtxWVScBJwObgaunmPvtwDPAZV0GzTJ3VT09nPNU4Ebgc0vLVfXyBLVPrAYX8H8UuCbJoUk2DWse+7yHz/kfgLuq6req6neAK4E3dZx+0a/194Azh9t4A3A0cMrI988EvjtB3Utf/9Fz3TPv2+u1j81Y91Q5MDLvbcC3q+qEqnon8GHg2AnrPxv4PHBOVT02ydgpvQR8MMnR89rgup02qaongEuBjw//Q9ZyFrC3qm4ajt0HXA5cnGTjFNPfA7yl42PnPffCVNWDwNeBTzH44bi5qvZ0GPpHwM+r6saRbe2squ9MUcYiXuvvMgxvBqH9IPB/SX4jySHAbwP3T1F7V7PsI63uX3Ope8IcWJr35WX75qNV9fmucyZ5F/BF4E86/jzMwyvAFxi8RnPxa/Pa0DSq6uHhkdIxwE/XeOgpwI5lY59L8hhwIvDDrnMm2QCcDXyp45C5zT2Fw5LsHFk+Erhjwm18FrgPeBnY2nHM21n2nKexqNe6qn6S5JXhn79n8stfGL8PPAv8sONR6Ojr/UhVdfpTfNq65zB2VrPsX3Ore4IcWJr3vq7bXsEhwO3AH1bVj2fYzjRuAH6Y5K/nsbF1De+hLr9tA6x0H/9q61eytKO+jcFO982O4+Yx97ReHP45PJgwuYjuAQxAVf0syVeB56vqpTnXt5r1eK2Xjr7PBK5hEN5nMgjv73Wc/zWv9wRmqbvV/WvedXfJgV8dlNwA/AGDo/EzOgz5OYP94RLgL6aZc1rDX243A58AXpx1e+t6tUmSE4B9wBNjHrqLZTtVksOB44Cuf/Ys7ajHAwfT8TzsnOZeb68Ov7raBbxzhvnW47VeOu/9uwxOm3yfwZF31/Pds5il7lb3r7nVPUEOLM17+tJCVV3G4K+7LR2nexW4ADgjyZWT1Dkn1zL4xbFp1g2tW3gn2cLgDZbra3x3rDuBjUkuHI7dAGwDtlfVC5PMW1XPMvjNd0WSgzoMmdvcDfkWcEiSP19akeSMJO+eZCMLfq2/C7wPeKaq9lXVM8ARDAL8nknqnsIsdbe6f82l7glzAAb75qFJPjaybqL3Bob1vQ/4SJJLJhk7q+F++TUGAT6TRYf3YUuXCAH/Avwzg3Oyaxr+p54HnJ9kN/AQsJfBFRATq6r7gQcYvEu90LlbMPKc3zO8BGwXcBXwkym2tajX+kcMrjL5/rJ1z1bVU5PWPYlZ6m51/5qx7qlyYGTeDwDvTvJIkh8AX2bwpvwk9T8DnAP8VZI/nWTsHGxjsK/OJLaElaT2eIelJDXI8JakBhnektQgw1uSGrSelwpe6tj9f26fcxtj13PuFseu59yz1v0LVbUuX8C9jt3/5/Y5tzG21bp9vab/8rSJJDWot+u8D03q19f43bCX4tBV2hkcf9o71tz2k089zZajj5qqrhbHrufcPuc2xq7n3OPGPnr/6j2q1soBWDsLXq+v1477dz5VVWNv9+8tvLdkQ31osrtWf+HGn/3nnKuRtF4+ummiVtuvcSBmQTYdsaOqxjYI87SJJDXI8JakBhnektSgsR/GkGQfg+5sBzH4KJ8vA9dW1ST9oSVJc9Tlk3R+8WkbSY4BbgHeSMcPDJUkzd9Ep01q8g8LlST1YOJz3lX18HDcMcu/l+TSJPcmuXdv7x+/J0kHrmnfsFzxqLuqvlBVW6tq61oX3kuSZjNxeE/4YaGSpB5MFN5TfFioJKkHXa42OSzJTn55qeBXgGt6rUqStKax4V1VGxZRiCSpuy5H3lM5/rR3cOPdd001dj0b2dhEZ3Fafa1nqXsWrT5nfy764e3xktQgw1uSGmR4S1KDDG9JapDhLUkN6hTeSY5NcnuS3Un2JLkuycF9FydJWtnY8B52D7wVuK2qTgJOBjYDV/dcmyRpFV2OvM8C9lbVTQBVtQ+4HLg4yXSfMCxJmkmX8D4F2DG6oqqeAx4DThxdP9oS9smnnp5flZKk1+gS3oEVm3P/yvrRlrBbjj5qHvVJklbQJbx3AVtHVyQ5HDgO2NNHUZKktXUJ7zuBjUkuBEiyAdgGbK+qF/osTpK0srHhPezbfR5wfpLdwEPAXuDKnmuTJK2iU1fBqnoceH/PtUiSOvIOS0lqUG/9vGcxS//fWXst23t4Mgdin+dW657Fgfic93ceeUtSgwxvSWqQ4S1JDTK8JalBnd6wTLIP+NHw8Y8Af1ZV/9tnYZKk1XU98n6xqk6tqrcDzwCX9ViTJGmMaU6b3AO8Zd6FSJK6myi8h31NzgbuWOX7toSVpAXoGt6HJdkJPA0cCXxzpQfZElaSFmOic97A8cDBeM5bktbVRKdNqupZ4BPAFUkO6qckSdI4E79hWVX3Aw8AH55/OZKkLrq2hN28bNn2sJK0jrzDUpIatF+2hF3PNqMHYovTWaxX+971fK1nbTs8rVaf84H4c7EIHnlLUoMMb0lqkOEtSQ0yvCWpQZ3CO8mxSW5PsjvJniTXJTm47+IkSSsbG95JAtwK3FZVJwEnA5uBq3uuTZK0ii5H3mcBe6vqJoCq2gdcDlycZGOfxUmSVtYlvE8BdoyuqKrngMeAE/soSpK0ti7hHaC6rLeftyQtRpfw3gVsHV2R5HDgOGDP6Hr7eUvSYnQJ7zuBjUkuhF98ms42YHtVvdBncZKklY0N76oq4Dzg/CS7gYeAvcCVPdcmSVpF15awjwO2gZWk/YR3WEpSgzI4KzJ/W08/re69+65etr2/sm2m+uT+dWDIpiN2VNXWcY/zyFuSGmR4S1KDDG9JapDhLUkNMrwlqUGdrvNOchSDOy0B3gzsA54cLv9eVb3cQ22SpFV0vUnnaeBUgCRXAc9X1d/0WJckaQ2eNpGkBs01vG0JK0mLMdfwtiWsJC2Gp00kqUGGtyQ1yPCWpAZ1ulRwVFVd1UMdkqQJTBzeWt0sbTdt96lx1mv/mnVu9cPTJpLUIMNbkhpkeEtSgwxvSWqQ4S1JDRob3kmeX7Z8UZLr+ytJkjSOR96S1CDDW5Ia1OUmncOS7BxZPhK4Y6UHJrkUuBTgrccdN3t1kqQVdTnyfrGqTl36Aj692gNtCStJi+FpE0lqkOEtSQ0yvCWpQWPfsKyqzcuWtwPbe6pHktSBR96S1CD7ee8n7NWsPvl//PrjkbckNcjwlqQGGd6S1CDDW5Ia1Cm8M3B3kveOrLsgyTf6K02StJpOV5tUVSX5KPD3Sf4V2ABcDZzTZ3GSpJV1vlSwqh5M8nXgU8Am4Oaq2tNbZZKkVU16nfdngfuAl4Gty79pS1hJWoyJ3rCsqp8BXwW+UlUvrfB9W8JK0gJMc7XJq8MvSdI68VJBSWqQ4S1JDZq4MVVVXdVDHZKkCXjkLUkNsiXsfmKWtq62+9Q4tg1+/fHIW5IaZHhLUoMMb0lqkOEtSQ0aG95Jnl+2fFGS6/srSZI0jkfektQgw1uSGtTlOu/DkuwcWT4SuGOlB9oSVpIWo8uR94tVderSF/Dp1R5oS1hJWgxPm0hSgwxvSWqQ4S1JDRr7hmVVbV62vB3Y3lM9kqQOPPKWpAbtly1h17M96nrNfSC23Gy1De6s7VWntZ77V6v/V69nHnlLUoMMb0lqkOEtSQ0yvCWpQYa3JDWoU3gnOTbJ7Ul2J9mT5LokB/ddnCRpZV0+jCHArcBtVXUScDKwGbi659okSavocuR9FrC3qm4CqKp9wOXAxUk29lmcJGllXcL7FGDH6Iqqeg54DDhxdH2SS5Pcm+TeJ596en5VSpJeo0t4B6gu6+3nLUmL0SW8dwFbR1ckORw4DtjTR1GSpLV1Ce87gY1JLgRIsgHYBmyvqhf6LE6StLKx4V1VBZwHnJ9kN/AQsBe4sufaJEmr6NRVsKoeB97fcy2SpI68w1KSGpTBWZH525IN9SGmuwzc/r/S64e9wCeTTUfsqKqt4x7nkbckNcjwlqQGGd6S1CDDW5IaNPZSwST7gB8BBwGvAF8Grq2qV3uuTZK0ii7Xeb9YVacCJDkGuAV4I/CZPguTJK1uotMmVfUEcCnw8WGfb0nSOpj4nHdVPTwcd8zy7422hN27YiNCSdI8TPuG5YpH3aMtYQ9d+SGSpDmYOLyTnADsA56YfzmSpC4mCu8kW4Abgeurr/vqJUljdbna5LAkO/nlpYJfAa7ptSpJ0prGhndVbVhEIZKk7rzDUpIa1FtL2K2nn1b33n1XL9uWdGA4ENvJ2hJWkl7HDG9JapDhLUkNMrwlqUGdwjvJm5LckuThJDuS3JPkvL6LkyStbGx4D7sH3gZ8u6pOqKp3Ah8Gpn8bWJI0ky53WJ4FvFxVNy6tqKpHgc/3VpUkaU1dTpucAtzXZWOjLWGffOrp2SqTJK1qmq6CNyR5IMm/Lf/eaEvYLUcfNZ8KJUm/okt47wJOX1qoqsuAs4EtfRUlSVpbl/D+FnBoko+NrNvYUz2SpA7Ghvewb/cHgHcneSTJDxh8gvyn+i5OkrSyLlebUFX/zeDyQEnSfsA7LCWpQZ2OvCVpPczS1nWWdrKzzr0IHnlLUoMMb0lqkOEtSQ0yvCWpQV1bwj4/8u9zk+xO8tb+ypIkrWWiq02SnM2gm+AfV9Vj/ZQkSRqnc3gneRfwReDcqtrTX0mSpHG6nvM+BLgd+EBV/Xi1B9kSVpIWo2t4/xz4HnDJWg+yJawkLUbX8H4VuAA4I8mVPdYjSeqg8znvqnohyfuA7yT5aVV9qce6JElrmOhqk6p6Jsk5wLeTPFVVt/dUlyRpDV1bwm4e+ffjwG/2VpEkaSzvsJSkBtkSVtLr0qwtXWdpKbuIdrIeeUtSgwxvSWqQ4S1JDTK8JalBU4X3aItYSdLieeQtSQ0yvCWpQXMNb1vCStJizDW8bQkrSYvhaRNJapDhLUkNMrwlqUFThfdoi1hJ0uJ55C1JDTK8JalB9vOWpBXM0pN7ll7gXXnkLUkNMrwlqUGGtyQ1yPCWpAaNDe8klWTbyPIVSa7qtSpJ0pq6HHm/BHwwydF9FyNJ6qZLeL8CfAG4vOdaJEkddT3nfQPwkSRvXOtB9vOWpMXoFN5V9RxwM/CJMY+zn7ckLcAkV5tcC1wCbOqpFklSR53Du6qeAb7GIMAlSeto0uu8twFedSJJ62xsY6rR3t1V9VNgY68VSZLG8g5LSWpQqqqfDSdPAo/2snFJev06vqq2jHtQb+EtSeqPp00kqUGGtyQ1yPCWpAYZ3pLUIMNbkhpkeEtSgwxvSWqQ4S1JDTK8JalB/w+gTHoVo1aQxgAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def plot_heatmap(dotplot_matrix, seq1=\"DOROTHYHODGKIN\", seq2=\"DOROTHYCROWFOOTHODGKIN\"):\n",
    "    heatmap = plt.imshow(dotplot_matrix, cmap=\"Reds\")\n",
    "    plt.gca().yaxis.set_major_locator(plt.MultipleLocator(1))\n",
    "    plt.gca().yaxis.set_ticklabels(\"_\" + seq1)\n",
    "    plt.gca().xaxis.set_major_locator(plt.MultipleLocator(1))\n",
    "    plt.gca().xaxis.set_ticklabels(\"_\" + seq2)\n",
    "    plt.tick_params(axis=\"x\", top=True, bottom=False, labeltop=True, labelbottom=False)\n",
    "    plt.show()\n",
    "    return \n",
    "\n",
    "plot_heatmap(dotplot_dh)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Task 2: Consensus and profiles from ungapped aligned DNA sequences\n",
    "1) Solve this Rosalind problem on creating consensus and profiles from an alignment of ungapped DNA sequences: https://rosalind.info/problems/cons/\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   0  1  2  3  4\n",
      "A  4  0  0  0  1\n",
      "C  0  2  2  0  0\n",
      "T  0  2  1  4  3\n",
      "G  0  0  1  0  0\n",
      "Consensus sequence is: ACCTT\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Profile\n",
    "def make_profile(path: str = \"data/consensus_and_pwm/consensus_test3.fa\", profile_width: int = 5) -> pd.DataFrame:\n",
    "    profile = pd.DataFrame({\"A\": [0] * profile_width,\n",
    "                            \"C\": [0] * profile_width,\n",
    "                            \"T\": [0] * profile_width,\n",
    "                            \"G\": [0] * profile_width})\n",
    "    with open(path, \"r\") as fasta:\n",
    "        for record in SeqIO.parse(fasta, \"fasta\"):\n",
    "            count = 0\n",
    "            for i in record.seq:\n",
    "                if i == \"A\":\n",
    "                    profile.at[count, 'A'] += 1\n",
    "                elif i == \"C\":\n",
    "                    profile.at[count, 'C'] += 1\n",
    "                elif i == \"T\":\n",
    "                    profile.at[count, 'T'] += 1\n",
    "                elif i == \"G\":\n",
    "                    profile.at[count, 'G'] += 1\n",
    "                count += 1\n",
    "    return profile   \n",
    "\n",
    "profile=make_profile()\n",
    "print(profile.T)\n",
    "\n",
    "# Consensus sequence\n",
    "print(\"Consensus sequence is:\", ''.join(profile.idxmax(axis=1).values))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2) Alter your code to return log-odds scores relative to a background probability for each DNA letter (0.25 by default)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          0         1         2         3    4\n",
      "A  1.321928 -1.000000 -1.000000 -1.000000  0.0\n",
      "C -1.000000  0.584963  0.584963 -1.000000 -1.0\n",
      "T -1.000000  0.584963  0.000000  1.321928  1.0\n",
      "G -1.000000 -1.000000  0.000000 -1.000000 -1.0\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "\n",
    "def PSSM (n, dflength: int = 4, pseudocount: int = 1):\n",
    "    return math.log2(((n+pseudocount)/(dflength+(pseudocount*4)))/0.25)\n",
    "\n",
    "profile_logodds = profile.T.map(PSSM)\n",
    "print(profile_logodds)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3) Use the BioPython class Bio.motifs to check the output is correct."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACCTT\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'A': [2.0, -inf, -inf, -inf, 0.0],\n",
       " 'C': [-inf, 1.0, 1.0, -inf, -inf],\n",
       " 'G': [-inf, -inf, 0.0, -inf, -inf],\n",
       " 'T': [-inf, 1.0, 0.0, 2.0, 1.5849625007211563]}"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import motifs\n",
    "\n",
    "read_motif = []\n",
    "with open(\"data/consensus_and_pwm/consensus_test3.fa\", \"r\") as fasta:\n",
    "    for record in SeqIO.parse(fasta, \"fasta\"):\n",
    "        read_motif.append(record.seq)\n",
    "\n",
    "# Consensus sequence\n",
    "motif_list = motifs.create(read_motif)\n",
    "print(motif_list.consensus)\n",
    "\n",
    "# PSSM (no pseudocounts, couldn't get this to work)\n",
    "motif_list.pssm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Task 3: Write a program to find all open reading frames in both strands of a DNA string given as input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Task 4: Needleman-Wunsch Algorithm\n",
    "\n",
    "Write a program that will solve the global sequence alignment problem for two sequences (start with DNA for simplicity) with linear gap penalties using the dynamic programming approach suggested by Needleman and Wunsch.\n",
    "\n",
    "Hint: You can use two matrices (2D arrays), one to keep information on the best wat to reach each element and one to use for trace-back"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
